### ISSUE TITLE
 Impact of gradient complexity (Galton Board)

### ISSUE THREAD

**Issue by Justinezgh** [Justinezgh/SBI-Diff-Simulator#31]
Gradients variance as a function of board length:
![image](https://user-images.githubusercontent.com/72011736/202703956-dd382f75-e2d5-4f6a-b136-487edcbe24c3.png)

code: 
```
var_grad = []

for i in tqdm(range(2, 302, 10)):

  model = partial(galton_board, n_rows = i, n_nails = 31)
  cond_model = condition(model, {'theta': -0.8, 
                                 'y':19})
  (log_probs, samples), scores = get_samples_and_scores(cond_model, 
                                                        jax.random.PRNGKey(0), 
                                                        batch_size=20000, 
                                                        score_type='density', 
                                                        thetas=None)
  
  var_grad.append(jnp.std(scores))
```

I'm not sure why the variance decreases at some point

**Comment by Justinezgh** [comment]
![image](https://user-images.githubusercontent.com/72011736/202700781-a9a3523c-15ae-438e-b3bb-08daa3fbc5ab.png)

**Comment by Justinezgh** [comment]
I inferered the posterior with 1000 simulations from different  Galton boards: 10 rows, 30 rows and 50 rows. 

![image](https://user-images.githubusercontent.com/72011736/202711503-0d4a7896-fbe9-4dd1-b1d2-7927b26eeb4b.png)

![image](https://user-images.githubusercontent.com/72011736/202711536-87855325-a311-4e70-838e-95b55a1879b6.png)

![image](https://user-images.githubusercontent.com/72011736/202711574-8ae598a2-f8bc-41a8-970a-670125ace379.png)


![image](https://user-images.githubusercontent.com/72011736/202711666-7501b525-9141-4948-97ab-ab5766e81501.png)

**Comment by Justinezgh** [comment]
Here are the score std for different Galton Board lengths (10, 20, 30, 40 and 50 rows), for each length I computed the std in each bins for 10 different thetas (-1, -0.9,...,0) that's why for each length there are 10 curves.

![image](https://user-images.githubusercontent.com/72011736/203550244-9d150fde-d109-431c-80f0-09854fa1430e.png)


And here are the sbi results. I trained 3 NFs for each Galton Board lengths and score weight, that's why there are 3 curves of the same color. And I used 80 simulations. 
![image](https://user-images.githubusercontent.com/72011736/203551261-7b5274e8-e703-4340-bfd0-58b64f3ffdcc.png)

![image](https://user-images.githubusercontent.com/72011736/203551323-8dd11698-62ef-434c-b442-c949101d5de6.png)

![image](https://user-images.githubusercontent.com/72011736/203551359-ddb1e68b-04ac-471f-91a1-1ca8c9b40571.png)

![image](https://user-images.githubusercontent.com/72011736/203551382-e0cce9dc-1d2a-40bb-9fda-3b5a12549838.png)

**Comment by Justinezgh** [comment]
In the ml4astro paper we found out that gradients help in the case of a narrow proposal distribution (which was also the prior distribution) :
<img width="363" alt="image" src="https://user-images.githubusercontent.com/72011736/205671928-81039915-2eb1-455f-bc12-2667ec0b2a1a.png">

So I computed the standard deviation of the score for LV with wide and narrow prior:
<img width="700" alt="image" src="https://user-images.githubusercontent.com/72011736/205673644-d616b472-e271-4107-9362-d0bf2df474b2.png">

1. can we say that this confirms that the higher the variance the less the score helps ? 
2. this is probably why sequential method does not work better with score

**Comment by Justinezgh** [comment]
I increased the number of nails in the GB (10000): 
```python 
def galton_board(y = None,  n_rows = 20, n_nails = 10000):
  """
  Probabilistic model for the Generalized Galton Board example as described
  in https://github.com/johannbrehmer/simulator-mining-example.

  Args:
    y: bin in which the ball ends up 
    n_rows: number of rows in the galton board
    n_nails: number of nails in the galton board
  """

  theta = numpyro.sample('theta', dist.Uniform(-1, 0))
  pos = n_nails // 2

  for level in range(n_rows):
    pos = numpyro.sample('z%d' %level,  
                      dist.TransformedDistribution(dist.Bernoulli(1 - nail_positions(theta, n_rows, n_nails, level, pos)), 
                                              dist.transforms.AffineTransform( - (level % 2) + pos, 1)))
  y = numpyro.sample("y", 
                      dist.TransformedDistribution(dist.Bernoulli(1 - nail_positions(theta, n_rows, n_nails, level, pos)), 
                                              dist.transforms.AffineTransform(- (level % 2) + pos, 1)), 
                      obs=y)

  return y 
```
![image](https://user-images.githubusercontent.com/72011736/208669456-631f9f21-d133-4ac0-b62b-922f574d060a.png)

and then rescaled the output distribution to have the same regardless of the number of rows:
```python 
def galton_board(y = None,  n_rows = 20, n_nails = 10000):

  theta = numpyro.sample('theta', dist.Uniform(-1, 0))
  pos = n_nails // 2

  for level in range(n_rows):
    pos = numpyro.sample('z%d' %level,  
                      dist.TransformedDistribution(dist.Bernoulli(1 - nail_positions(theta, n_rows, n_nails, level, pos)), 
                                              dist.transforms.AffineTransform( - (level % 2) + pos, 1)))
  
  y = numpyro.sample("y", dist.Normal((pos - n_nails // 2) * jnp.sqrt(3.9/n_rows), 0.5))

  return y 
```
![image](https://user-images.githubusercontent.com/72011736/208668986-20b6cc9d-3d9d-49e6-9d22-d0aba8f739ca.png)

But even if the std of the score increases with the number of rows 
![image](https://user-images.githubusercontent.com/72011736/208670307-97fd73e7-b7aa-4aa5-ae70-2b416f7c1597.png)
the snr remains the same 
![image](https://user-images.githubusercontent.com/72011736/208670361-b1183d6c-c2c7-4460-b4ff-dfe9f43a57f1.png)
and the inference results are the same 
![image](https://user-images.githubusercontent.com/72011736/208678865-162ec3c5-e075-461d-91dd-d59c08dc9e45.png)
![image](https://user-images.githubusercontent.com/72011736/208678920-fee2d986-3ca8-4fbe-a4ed-1ab9a98f32ff.png)

(you can find the notebook [here](https://github.com/Justinezgh/SBI-Diff-Simulator/blob/u/Justinezgh/gradients_stochasticity/notebooks/dev/GradientsStochasticity_GB.ipynb))

**Comment by Justinezgh** [comment]
I changed the Galton board task to have the snr decreasing with the nb of rows
 
![image](https://user-images.githubusercontent.com/72011736/210791460-443d648e-60ce-43d7-aff8-701321ea05c2.png)

![image](https://user-images.githubusercontent.com/72011736/210791498-4a314ac7-5f9f-4f4d-be0b-abcf010d33a9.png)

![image](https://user-images.githubusercontent.com/72011736/210791516-7dd22c42-524a-48da-83d1-511eb1277e4e.png)

![image](https://user-images.githubusercontent.com/72011736/210791604-98911346-c904-43d9-b309-d1bbcd43fecc.png)

![image](https://user-images.githubusercontent.com/72011736/210791621-cf67bfe0-a56e-4af1-9ae4-82625847d3f4.png)

**Comment by Justinezgh** [comment]
I re designed the Galton Board to have a simulator that verifies the following requirements:
- the posterior p( theta | x ) should remain the same for each number of latent variables
- the SNR should decrease with the number of latent variables
- the likelihood p( x | theta ) should depend on theta

And I perform sbi not only on one simulation but on 20 simulations x that I summarize in an histogram of 21 bins. To have the corresponding score I compute sum(grad_{theta}p(theta|x_i, z_i)). But when I trained the nf on (theta, histogram, sum score) using the negative log likelihood loss + 0.01 * score matching loss I have this weird behavior: 
![image](https://user-images.githubusercontent.com/72011736/213457979-bd34e5a6-1cf5-4451-9b30-e5bd8b0bd9a7.png) 
and the results: 
![image](https://user-images.githubusercontent.com/72011736/213459283-89235abe-153d-489d-baaa-aa141970b536.png)
compared to those obtained only by training with the nll loss:
![image](https://user-images.githubusercontent.com/72011736/213459515-3d18beea-d1ac-448d-9859-09e6351df12d.png)

which gives me the impression that I'm not comparing the same kind of gradients. 
The [notebook](https://colab.research.google.com/drive/1-0N7ZFNPaKlbmWTOgRasxGDCsAQRgGDB?usp=sharing).
